torchtext/datasets/language_modeling.py

```diff
@@ -1,217 +1,236 @@
-                text field. The word vectors are accessible as
-                train.dataset.fields['text'].vocab.vectors.
-            Remaining keyword arguments: Passed to the splits method.
         """
-        TEXT = data.Field()
-
-        train, val, test = cls.splits(TEXT, root=root, **kwargs)
-
-        TEXT.build_vocab(train, vectors=vectors)
-
-        return data.BPTTIterator.splits(
-            (train, val, test), batch_size=batch_size, bptt_len=bptt_len,
-            device=device)
 
+        super(LanguageModelingDataset, self).__init__()
+        self._data = data
+        self._vocab = vocab
+
+    def __getitem__(self, i):
+        return self._data[i]
+
+    def __len__(self):
+        return len(self._data)
+
+    def __iter__(self):
+        for x in self._data:
+            yield x
+
+    def get_vocab(self):
+        return self._vocab
+
+
+def _get_datafile_path(key, extracted_files):
+    for fname in extracted_files:
+        if key in fname:
+            return fname
+
+
+def _setup_datasets(dataset_name, tokenizer=get_tokenizer("basic_english"),
+                    root='.data', vocab=None, removed_tokens=[],
+                    data_select=('train', 'test', 'valid')):
+
+    if isinstance(data_select, str):
+        data_select = [data_select]
+    for item in data_select:
+        if item not in ('train', 'test', 'valid'):
+            raise TypeError('{} in data_select is not supported!'.format(item))
+
+    if dataset_name == 'PennTreebank':
+        extracted_files = []
+        if 'train' in data_select:
+            extracted_files.append(download_from_url(URLS['PennTreebank'][0], root=root))
+        if 'test' in data_select:
+            extracted_files.append(download_from_url(URLS['PennTreebank'][1], root=root))
+        if 'valid' in data_select:
+            extracted_files.append(download_from_url(URLS['PennTreebank'][2], root=root))
+    else:
+        dataset_tar = download_from_url(URLS[dataset_name], root=root)
+        extracted_files = [os.path.join(root, d) for d in extract_archive(dataset_tar)]
+
+    _path = {}
+    for item in data_select:
+        _path[item] = _get_datafile_path(item, extracted_files)
+
+    if vocab is None:
+        if 'train' not in _path.keys():
+            raise TypeError("Must pass a vocab if train is not selected.")
+        logging.info('Building Vocab based on {}'.format(_path['train']))
+        vocab = build_vocab_from_iterator(read_text_iterator(_path['train'], tokenizer))
+        logging.info('Vocab has {} entries'.format(len(vocab)))
+    else:
+        if not isinstance(vocab, Vocab):
+            raise TypeError("Passed vocabulary is not of type Vocab")
+
+    _data = {}
+    for item in _path.keys():
+        _data[item] = []
+        logging.info('Creating {} data'.format(item))
+        _iter = create_data_from_iterator(
+            vocab, read_text_iterator(_path[item], tokenizer), removed_tokens)
+        for tokens in _iter:
```

Comment:
List expressions here could help performance

```data[item] = list(tokens for tokens in iter)```.

But actually, there is an opportunity for a performance optimization if we support returning an IterableDataset. So instead of materializing the entire dataset we just initialize it with an iterator. I think this is worth thinking about because it could help us make large datasets easily available (especially using the background iterator). However, should we include this here at this time? It could be a simple flag such as "return_iterable_datasets".

cc @vincentqb

Reply:
I think for now, we should return the materialized token ids for consistency.

Reply:
This does seem like a good opportunity for optimization, and to start moving datasets in that direction. The return_iterable_datasets flag could default to "False" with a deprecation warning.

Do we know the length of `_iter`? If so, it's easy to return an iterable class that also offers a `__getitem__` that materialize lazyly.

The reason I'd see not to do this is if `_setup_datasets` is meant for users that would only experiment at small scale, or if we are moving away from this interface all together. Neither seems to be the case. I agree with consistency, but that would mean moving as many datasets as possible in that direction.

Thoughts?

Reply:
IterableDatasets are a pretty new addition [about 5 months ago](https://github.com/pytorch/pytorch/pull/19228). This could be an opportunity to embrace that. 

The length will be hard to know for large datasets. We could call into stat and use the filesize, but that won't necessarily get us the number of lines of text.

Materializing lazily is only possible if we can read at an offset, but is an interesting idea. A diskcache iterator could allow people to write out the token ids onto disk as it's being consumed, but I think that's something to throw on top (should be easy given that IterableDataset behaves like an iterator).

Reply:
I could write the iterabledataset and return the materialized one as default. Coming soon.